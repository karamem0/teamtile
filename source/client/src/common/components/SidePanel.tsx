//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { Event, EventHandler } from '../../types/Event';
import Presenter from './SidePanel.presenter';

interface SidePanelRenderProps {
  onOpenChange?: EventHandler<boolean>
}

interface SidePanelProps {
  content?: React.ReactNode,
  loading?: boolean,
  renderer?: (props: SidePanelRenderProps) => React.ReactNode,
  title?: React.ReactNode,
  onOpenChange?: EventHandler<boolean>
}

function SidePanel(props: Readonly<SidePanelProps>) {

  const {
    content,
    loading,
    renderer,
    title,
    onOpenChange
  } = props;

  const [ open, setOpen ] = React.useState<boolean>(false);

  const handleOpenChange = React.useCallback((event: Event, data?: boolean) => {
    setOpen(data ?? false);
    onOpenChange?.(event, data);
  }, [
    onOpenChange
  ]);

  return (
    <Presenter
      content={content}
      loading={loading}
      open={open}
      title={title}
      onOpenChange={handleOpenChange}>
      {
        renderer?.({ onOpenChange: handleOpenChange })
      }
    </Presenter>
  );

}

export default SidePanel;
