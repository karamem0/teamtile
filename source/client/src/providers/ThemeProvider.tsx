//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { Provider, themeNames } from '@fluentui/react-teams';

import { app } from '@microsoft/teams-js';

import { inTeams } from '../utils/Teams';

interface ThemeProviderProps {
  children?: React.ReactNode
}

function ThemeProvider(props: ThemeProviderProps) {

  const { children } = props;

  const [ lang, setLang ] = React.useState<string>('en-US');
  const [ themeName, setThemeName ] = React.useState<themeNames>(themeNames.Default);

  const handleThemeChange = (value: string) => {
    switch (value) {
      case 'dark':
        setThemeName(themeNames.Dark);
        break;
      case 'contrast':
        setThemeName(themeNames.HighContrast);
        break;
      default:
        setThemeName(themeNames.Default);
        break;
    }
  };

  React.useEffect(() => {
    if (!inTeams()) {
      return;
    }
    (async () => {
      await app.initialize();
      const context = await app.getContext();
      setLang(context.app.locale);
      handleThemeChange(context.app.theme);
      app.registerOnThemeChangeHandler(handleThemeChange);
    })();
  }, []);

  return (
    <Provider
      lang={lang}
      themeName={themeName}>
      {children}
    </Provider>
  );

}

export default ThemeProvider;
