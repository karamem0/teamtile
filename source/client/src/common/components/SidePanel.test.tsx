//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import SidePanel from './SidePanel.presenter';
import ThemeProvider from '../../providers/ThemeProvider';
import { render } from '@testing-library/react';

it('should create shapshot', async () => {
  const params = {
    open: true,
    content: (
      <div data-testid="Content" />
    ),
    title: (
      <div data-testid="Title" />
    )
  };
  const { asFragment } = render(
    <ThemeProvider>
      <SidePanel {...params} />
    </ThemeProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
