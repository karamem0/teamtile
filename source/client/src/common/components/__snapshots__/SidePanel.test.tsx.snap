// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a shapshot when the loading parameter is false 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="rmnljus fui-OverlayDrawer__backdrop ___mgf8q40_1ccxybz f1euv43f"
  />
  .emotion-0 {
  margin-bottom: 1rem;
}

<aside
    aria-labelledby="dialog-title-r6"
    aria-modal="true"
    class="fui-OverlayDrawer r7dfyj8 ___12p5wfa_jv28t00 fsqykmd f1e31b4d frppm18 f1exhnwo f1euv43f"
    data-tabster="{"restorer":{"type":0},"modalizer":{"id":"modal-r5","isOthersAccessible":false,"isTrapped":true}}"
    role="dialog"
    tabindex="-1"
  >
    <header
      class="fui-DrawerHeader r3ovn4i"
      role="none"
    >
      <div
        class="fui-DrawerHeaderTitle ___1g6dz7n_17w6zju f22iagw f1869bpl f122n59 fsnqrgy"
      >
        <h2
          class="fui-DrawerHeaderTitle__heading fui-DialogTitle rxjm636"
          id="dialog-title-r6"
        >
          <div
            data-testid="test-Title"
          />
        </h2>
        <div
          class="fui-DrawerHeaderTitle__action fui-DialogTitle__action r13kcrze ___zdrx2v0_7c5v900 faqnl2i"
        >
          <button
            aria-label="Close"
            class="fui-Button r1alrhcs ___w3o4yv0_df8ghi0 fhovq9v f1p3nwhy f11589ue f1q5o8ev f1pdflbu fkfq4zb f1t94bn6 f1s2uweq fr80ssc f1ukrpxl fecsdlb fnwyq0v ft1hn21 fuxngvv fy5bs14 fsv2rcd f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1omzyqd f1dfjoow f1j98vj9 fj8yq94 f4xjyn1 f1et0tmh f9ddjv3 f1wi8ngl f18ktai2 fwbmr0d f44c6la"
            type="button"
          >
            <span
              class="fui-Button__icon rywnvv2"
            >
              <svg
                aria-hidden="true"
                class="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m2.59 2.72.06-.07a.5.5 0 0 1 .63-.06l.07.06L8 7.29l4.65-4.64a.5.5 0 0 1 .7.7L8.71 8l4.64 4.65c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L8 8.71l-4.65 4.64a.5.5 0 0 1-.7-.7L7.29 8 2.65 3.35a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </header>
    <div
      class="fui-DrawerBody ri6rvx7 emotion-0"
      role="group"
      tabindex="0"
    >
      <div
        data-testid="test-Content"
      />
    </div>
  </aside>
  .emotion-0 {
  min-height: 100vh;
  line-height: 1.25em;
  background-color: #f5f5f5;
}

<div
    class="fui-FluentProvider fui-FluentProviderr4 ___jdtuxv0_17k0bs4 f19n0e5 fxugw4r f1o700av fk6fouc fkhj508 figsok6 f1i3iumi"
    dir="ltr"
  >
    <div
      class="emotion-0"
    >
      <span
        hidden=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`should create a shapshot when the loading parameter is true 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="rmnljus fui-OverlayDrawer__backdrop ___mgf8q40_1ccxybz f1euv43f"
  />
  .emotion-0 {
  margin-bottom: 1rem;
}

<aside
    aria-labelledby="dialog-title-r2"
    aria-modal="true"
    class="fui-OverlayDrawer r7dfyj8 ___12p5wfa_jv28t00 fsqykmd f1e31b4d frppm18 f1exhnwo f1euv43f"
    data-tabster="{"restorer":{"type":0},"modalizer":{"id":"modal-r1","isOthersAccessible":false,"isTrapped":true}}"
    role="dialog"
    tabindex="-1"
  >
    <header
      class="fui-DrawerHeader r3ovn4i"
      role="none"
    >
      <div
        class="fui-DrawerHeaderTitle ___1g6dz7n_17w6zju f22iagw f1869bpl f122n59 fsnqrgy"
      >
        <h2
          class="fui-DrawerHeaderTitle__heading fui-DialogTitle rxjm636"
          id="dialog-title-r2"
        >
          <div
            data-testid="test-Title"
          />
        </h2>
        <div
          class="fui-DrawerHeaderTitle__action fui-DialogTitle__action r13kcrze ___zdrx2v0_7c5v900 faqnl2i"
        >
          <button
            aria-label="Close"
            class="fui-Button r1alrhcs ___w3o4yv0_df8ghi0 fhovq9v f1p3nwhy f11589ue f1q5o8ev f1pdflbu fkfq4zb f1t94bn6 f1s2uweq fr80ssc f1ukrpxl fecsdlb fnwyq0v ft1hn21 fuxngvv fy5bs14 fsv2rcd f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1omzyqd f1dfjoow f1j98vj9 fj8yq94 f4xjyn1 f1et0tmh f9ddjv3 f1wi8ngl f18ktai2 fwbmr0d f44c6la"
            type="button"
          >
            <span
              class="fui-Button__icon rywnvv2"
            >
              <svg
                aria-hidden="true"
                class="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m2.59 2.72.06-.07a.5.5 0 0 1 .63-.06l.07.06L8 7.29l4.65-4.64a.5.5 0 0 1 .7.7L8.71 8l4.64 4.65c.18.17.2.44.06.63l-.06.07a.5.5 0 0 1-.63.06l-.07-.06L8 8.71l-4.65 4.64a.5.5 0 0 1-.7-.7L7.29 8 2.65 3.35a.5.5 0 0 1-.06-.63l.06-.07-.06.07Z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </header>
    <div
      class="fui-DrawerBody ri6rvx7 emotion-0"
      role="group"
      tabindex="0"
    >
      <div
        class="fui-Spinner r82apo5"
        role="progressbar"
      >
        <span
          class="fui-Spinner__spinner rvgcg50 ___1xxu43e_eahmz60 f1d2rq10 f1szoe96 fb52u90"
        >
          <span
            class="fui-Spinner__spinnerTail rxov3xa"
          />
        </span>
      </div>
    </div>
  </aside>
  .emotion-0 {
  min-height: 100vh;
  line-height: 1.25em;
  background-color: #f5f5f5;
}

<div
    class="fui-FluentProvider fui-FluentProviderr0 ___jdtuxv0_17k0bs4 f19n0e5 fxugw4r f1o700av fk6fouc fkhj508 figsok6 f1i3iumi"
    dir="ltr"
  >
    <div
      class="emotion-0"
    >
      <span
        hidden=""
      />
    </div>
  </div>
</DocumentFragment>
`;
